### ISSUE TITLE
Improve GroupAdjacent

### ISSUE THREAD

**Issue by jods4** [morelinq/MoreLINQ#1032]
### Problem
`GroupAdjacent` doesn't let you cheaply compare a small key (e.g. single field) to define groups, then use different selectors for the group key and elements.

For example, consider an `IEnumerable` with the following fields:
```
aKey
a1
a2
a3
b1
b2
b3
```

You want to efficiently group by `aKey`, then use `x => A(x.AKey, x.a1, x.a2, x.a3)` as group key and `x => B(x.b1, x.b2, x.b3)` as group elements.

One option that works today is `GroupAdjacent(x => A(...), x => B(...))` and have a comparer or override `Equals` on `A`.
The drawback is that one `A` instance is generated for every element in the sequence, when you only need a single one per group.

### Proposed solution 1
A new overload? 
`GroupAdjacent(F<X, A> keySelector, F<X, B> elementSelector, F<X, Y> equalitySelector)`

### Proposed solution 2
`GroupAdjacent` doesn't stream its results, it actually buffers each group in a list. Internally these are `IList` but only exposed as `IEnumerable`.
Exposing the `IList` would allow the following solution: 
```cs
seq.GroupAdjacent(
  keySelector: x => x.aKey,
  resultSelector: (key, IList elements) => (A(elements[0]), elements.Select(B))
)  
```
Today, this is doable with a cast from `IEnumerable` to `IList` but it can break with any MoreLINQ release as it relies on an implementation detail.
Alternatively, using `elements.First()` instead of `elements[0]` will always work but can be quite dangerous if the implementation of `GroupAdjacent` becomes non-buffering, as `elements` is enumerated twice.

### What about non-buffering?
When working with long groups, it'd be interesting to not buffer the groups.
Maybe a new API `StreamGroupAdjacent` that has an added constraint that group elements must be consumed in order and only once (but not necessarily fully).

### Predicate vs IEqualityComparer 
It'd also be nice to have overloads accepting `Func<X, X, bool>` instead of `IEqualityComparer`. 
`GroupAdjacent` is relies on adjacency, so uses only equality and never hashes.

**Comment by atifaziz** [comment]
Sorry, but I am having trouble to understand the problem and what you're asking for.

If this is the input:

    aKey
    a1
    a2
    a3
    b1
    b2
    b3

What do you expect the output to be?

**Comment by jods4** [comment]
Basically, I added this to my project. It's a slightly less efficient of doing Proposed Solution 1:
```csharp
IEnumerable<IGrouping<G, E>> GroupAdjacent<T, K, G, E>(
  this IEnumerable<T> source,
  Func<T, K> key,
  Func<T, G> group,
  Func<T, E> element)
{
  K cachedKey = default;
  G cachedGroup = default;
  return source.GroupAdjacent(
    x => 
    {
      var nextKey = key(x);
      if (nextKey != cachedKey || cachedGroup is null)
        (cachedKey, cachedGroup) = (nextKey, group(x));
      return cachedKey;
    },
    element
}
```

The key point is that I don't want to create a `G` instance for every item in the source enumeration, only once per group.

**Comment by atifaziz** [comment]
Thanks for sharing the code. It helps to understand the mechanics a bit. So if the grouped elements were exposed as a list instead of a sequence, like your proposed solution 2, then it would solve your problem?

`GroupAdjacent` signatures were modeled after `GroupBy`, which is why the group elements are typed as `IEnumerable<>`.

**Comment by atifaziz** [comment]
Thanks for the clarifications.

I think [we've been down this road before, meaning that what you want here is (more generally speaking) `AggregateAdjacent`](https://github.com/morelinq/MoreLINQ/issues/836#issuecomment-1086006199)?

**Comment by jods4** [comment]
ah, I had forgotten this :)

What I want can be implemented on top of `AggregateAdjacent`, if that's what you mean:
```cs
IEnumerable<IGrouping<G, E>> GroupByAdjacent<T, K, G, E>(
  this IEnumerable<T> source,
  Func<T, K> key,
  Func<T, G> group,
  Func<T, E> element)
{
  return source.AggregateAdjacent(
    key,
    (key, x) => new Grouping(group(x)),
    (key, g, x) => 
    { 
      g.Add(element(x)); 
      return g; 
    },
    (key, g, list) => g
  );
}
```

**Comment by atifaziz** [comment]
@jods4 Yes, that's it!

**Comment by viceroypenguin** [comment]
What about the following with the existing api:
```cs
seq.GroupAdjacent(
  x => (x.aKey, x.a1, x.a2, x.a3),
  x => new B(x.b1, x.b2, x.b3),
  (k, g) => (new A(k.aKey, k.a1, k.a2, k.a3), g));
```

Unless `a1

**Comment by jods4** [comment]
@viceroypenguin 
Minor point: I have more `a_n` properties than I want to copy each iteration (ValueTuple is a struct)...
Major point: MoreLinq would then compare all `a_n` (default struct comparison) for each element, I want to compare `aKey` only.
